agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java

```diff
@@ -75,12 +102,32 @@ protected URLConnection openConnection(final URL url) throws IOException {
       // nullInputStream() is not available until Java 11
       return new InternalJarURLConnection(url, new ByteArrayInputStream(new byte[0]));
     }
-    final JarEntry entry = filenameToEntry.get(filename);
-    if (null != entry) {
-      return new InternalJarURLConnection(url, bootstrapJarFile.getInputStream(entry));
+    // believe it or not, we're going to get called twice for this,
+    // and the key will be a new object each time.
+    Pair<String, JarEntry> pair = cache.get();
+    if (null == pair || !filename.equals(pair.getLeft())) {
+      StringBuilder sb = JAR_ENTRY_QUERY.get();
+      sb.append(this.name).append(filename);
+      if (filename.endsWith(".class")) {
+        sb.append("data");
+      }
+      String classFileName = sb.toString();
+      sb.setLength(0);
+      JarEntry entry = bootstrapJarFile.getJarEntry(classFileName);
+      if (null != entry) {
+        pair = Pair.of(filename, entry);
+        // this mechanism intentionally does not ensure visibility of this write, because it doesn't
```

Comment:
This is because this is a performance optimization, not for correctness, I'm presuming

Reply:
1. Reference assignment is atomic so this cannot lead to inconsistency. 
2. If some other thread doesn't see the write, they just see another instance or null, and may or may not do extra work.

Reply:
> may or may not do extra work.

So the answer is yes :) I think the comment can be more self-explantory by talking about the motivation instead of the detail

`// This is a performance optimization, so in the rare event this write is not visible to another thread, it just means the same work is recomputed but does not affect consistency.`

Reply:
updated comment